src/ctap/apdu.rs

```diff
@@ -0,0 +1,303 @@
+use alloc::vec::Vec;
+use core::convert::TryFrom;
+
+type ByteArray = &'static [u8];
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(non_camel_case_types)]
+#[derive(PartialEq)]
+pub enum ApduStatusCode {
+    SW_SUCCESS,
+    /// Command successfully executed; 'XX' bytes of data are
+    /// available and can be requested using GET RESPONSE.
+    SW_GET_RESPONSE,
+    SW_WRONG_DATA,
+    SW_WRONG_LENGTH,
+    SW_COND_USE_NOT_SATISFIED,
+    SW_FILE_NOT_FOUND,
+    SW_INCORRECT_P1P2,
+    /// Instruction code not supported or invalid
+    SW_INS_INVALID,
+    SW_CLA_INVALID,
+    SW_INTERNAL_EXCEPTION,
+}
+
+impl From<ApduStatusCode> for ByteArray {
+    fn from(status_code: ApduStatusCode) -> ByteArray {
+        match status_code {
+            ApduStatusCode::SW_SUCCESS => b"\x90\x00",
+            ApduStatusCode::SW_GET_RESPONSE => b"\x61\x00",
+            ApduStatusCode::SW_WRONG_DATA => b"\x6A\x80",
+            ApduStatusCode::SW_WRONG_LENGTH => b"\x67\x00",
+            ApduStatusCode::SW_COND_USE_NOT_SATISFIED => b"\x69\x85",
+            ApduStatusCode::SW_FILE_NOT_FOUND => b"\x6a\x82",
+            ApduStatusCode::SW_INCORRECT_P1P2 => b"\x6a\x86",
+            ApduStatusCode::SW_INS_INVALID => b"\x6d\x00",
+            ApduStatusCode::SW_CLA_INVALID => b"\x6e\x00",
+            ApduStatusCode::SW_INTERNAL_EXCEPTION => b"\x6f\x00",
+        }
+    }
+}
+
+#[allow(non_camel_case_types, dead_code)]
+pub enum ApduIns {
+    SELECT = 0xA4,
+    READ_BINARY = 0xB0,
+    GET_RESPONSE = 0xC0,
+}
+
+#[cfg_attr(test, derive(Clone, Debug))]
+#[allow(dead_code)]
+#[derive(Default, PartialEq)]
+pub struct ApduHeader {
+    cla: u8,
+    ins: u8,
+    p1: u8,
+    p2: u8,
+}
+
+impl From<&[u8]> for ApduHeader {
```

Comment:
Considering the header has a fixed size, shouldn't we enforce it instead of using a slice?
Otherwise I guess the app will panic if the slice is less than 4 bytes long.

Reply:
Enforcing the header length results in this
```
   --> src/ctap/apdu.rs:122:28
    |
122 |             header: header.into(),
    |                            ^^^^ the trait `std::convert::From<&[u8]>` is not implemented for `ctap::apdu::ApduHeader`
```

Reply:
@kaczmarczyck this is also something that you commented on.

Reply:
Yes, you'd have to call it with an array. `array_ref!` could help.